backend/src/email/middlewares/email-template.middleware.ts

```diff
@@ -155,58 +149,111 @@ const uploadCompleteHandler = async (
+      throw new InvalidRecipientError()
+    }
 
-      const recipientCount = records.length
+    // Store temp filename
+    await TemplateService.storeS3TempFilename(+campaignId, filename)
 
-      await updateCampaignAndMessages(s3Key, campaignId, filename, records)
+    try {
+      // Return early because bulk insert is slow
+      res.sendStatus(202)
 
-      if (!res.headersSent) {
-        return res.json({
-          num_recipients: recipientCount,
-          preview: {
-            ...hydratedRecord,
-            // eslint-disable-next-line @typescript-eslint/camelcase
-            reply_to: emailTemplate.replyTo || null,
-          },
-        })
-      }
+      // Slow bulk insert
+      await updateCampaignAndMessages(+campaignId, s3Key, filename, records)
     } catch (err) {
+      // Do not return any response since it has already been sent
       logger.error(
-        `Error parsing file for campaign ${campaignId}. ${err.stack}`
+        `Error storing messages for campaign ${campaignId}. ${err.stack}`
       )
-      throw err
+      // Store error to return on poll
+      TemplateService.storeS3Error(+campaignId, err.message)
     }
   } catch (err) {
-    if (!res.headersSent) {
-      const userErrors = [
-        RecipientColumnMissing,
-        MissingTemplateKeysError,
-        InvalidRecipientError,
-        UnexpectedDoubleQuoteError,
-      ]
-
-      if (userErrors.some((errType) => err instanceof errType)) {
-        return res.status(400).json({ message: err.message })
-      }
+    const userErrors = [
+      RecipientColumnMissing,
+      MissingTemplateKeysError,
+      InvalidRecipientError,
+      UnexpectedDoubleQuoteError,
+    ]
+
+    if (userErrors.some((errType) => err instanceof errType)) {
+      return res.status(400).json({ message: err.message })
     }
+    return next(err)
+  }
+}
+
+/*
+ * Returns status of csv processing
+ */
+const pollCsvStatusHandler = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<Response | void> => {
+  try {
+    const { campaignId } = req.params
+    const {
+      isCsvProcessing,
+      filename,
+      tempFilename,
+      error,
+    } = await TemplateService.getCsvStatus(+campaignId)
+
+    // If done processing, returns num recipients and preview msg
+    let numRecipients, preview
+    if (!isCsvProcessing) {
+      ;[numRecipients, preview] = await Promise.all([
+        EmailMessage.count({
+          where: { campaignId: +campaignId },
+        }),
+        EmailService.getHydratedMessage(+campaignId),
+      ])
+    }
+
+    res.json({
+      is_csv_processing: isCsvProcessing,
+      csv_filename: filename,
+      temp_csv_filename: tempFilename,
+      csv_error: error,
+      num_recipients: numRecipients,
+      preview,
+    })
+  } catch (err) {
+    next(err)
+  }
+}
+
+/*
+ * Deletes csv error and temp csv name from db
+ */
+const deleteCsvErrorHandler = async (
```

Comment:
temp_filename exists --> processing state
temp_filename + error exist --> upload fails, rolled back

So if the previous csv upload was successful and the current one has failed, it would have rolled back to the previous successful state. The error string and the temp file name is there to provide user with the error info when he returns to the page. So removing those keys does not change the fact that it has rolled back to its previous state. On the other hand, deleting the s3_object will remove the previous successful csv keys.